etc/statistic_summary_single_column.c

```diff
@@ -0,0 +1,316 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <errno.h>
+#include <limits.h>
+
+#define INPUT_BUFFER 256
+#define REALOC_GROW 64
+
+/*
+   Takes a stream from stdin with small integers (one per line) and print some basic statistics
+   that some might find useful. A bit like the 'summary()' command in R with the addition
+   of the standard deviation and the first and third quantile.
+   It has been written for a single use case with known input and not for general usage.
+*/
+
+/*
+   Simplified sqrt within 1 ulp, no rounding; good enough to compute the SD
+
+   From a copy of a drafted paper by Prof W. Kahan and K.C. Ng, written in May, 1986.
+   That copy can be found in the file https://www.netlib.org/fdlibm/e_sqrt.c
+*/
+static double s_sqrt(double d)
```

Comment:
Why not libm sqrt?

Reply:
No libm in LTM.

Reply:
You are also using libc printf. This is the tuner and not the lib. Please just use libc for simplicity.
For the lib, I totally agree, there shouldn't be libm and only a bit of libc.

Reply:
> This is the tuner and not the lib

It is automated now and part of the buildprocess, so the same restrictions as for LTM itself apply.

Using floating point is already questionable here because I only use the medians as the cut-off values, the rest of the statistics is only there to approximate the error which isn't even used yet for evaluation directly, only for manual inspection (answering the question: Can I trust the result?).

I am a bit torn here: throw it all out, keep it as it is (with floating point) or redo the float part with some simple fixed-point math to have both. But the last one has definitely the largest amount of work involved.

Reply:
What is automated? Please don't replicate libc functionality of it is not strictly necessary. This will only be a maintanance burden. And I am only talking about tools.

Within the lib it is perfectly reasonable to require only a libc subset which is good for embedding.

If restrictions apply they should be lifted.

Reply:
> What is automated?

The evaluation of the TC cut-off points? That's why you called it "autotune"?

Decided to strip all function not necessary to do the evaluation of the TC cut-off points, so no floats anymore. Dependencies are now down to `strtol`, `fgets` and `(f)printf`. which even the most barebone libc should offer.

Tuning must happen on the target-architecture, where, for example in the case of small MCUs not everything what gets offered by their (the IDEs of the manufacturers) libc is, how shall I put it, usable in all and every circumstance, so we should reduce the number of dependencies. Not at all cost, of course, but if it is cheap to have like here, we should do it.

Be conservative in what you do, be liberal in what you accept from others. (Jon Postel)

I was also thinking about exchanging heap for stack to get rid of `malloc` but decided against it, didn't want to implement a sliding-window median.

Reply:
> Decided to strip all function not necessary to do the evaluation of the TC cut-off points, so no floats anymore. Dependencies are now down to strtol, fgets and (f)printf. which even the most barebone libc should offer.

fprintf/printf usually does floating point. A barebone libc/libm usually provides sqrt.

> Tuning must happen on the target-architecture, where, for example in the case of small MCUs not everything what gets offered by their (the IDEs of the manufacturers) libc is, how shall I put it, usable in all and every circumstance, so we should reduce the number of dependencies. Not at all cost, of course, but if it is cheap to have like here, we should do it.

I agree that it should be done on the target platform. But you are prematurely assuming certain things, which adds unnecessary code here, which I am against. You assume:

- The presence of floating point in IEEE format
- The presence of fprintf
- The presence of malloc
- That sqrt does not exist

This does not make sense.

I wrote bare metal code before. And the first things I got rid were malloc and fprintf. For sqrt you can just link against a libm. This is not our business. Furthermore the version you added accesses the floating point internals which is not portable and worse than relying on the target libm.

Reply:
Just to make it clear - I very much appreciate the efforts to only rely on a minimal amount of libc. I am using tommath in a setting without one actually. But we should stay reasonable and code maintainability trumps without-libc-ness. In particular if it is about tooling. The library code is a harder story.

I think the following is overkill for tooling:

* Not relying on libm for tooling
* Fixed point math
* Avoiding malloc or printf

I will go through the list of the unwanted functions.

Reply:
The following are currently avoided in the lib:

* memcpy, memset, memmove, memcmp: I think these should not be avoided even in the lib since every target platform supports those and they will compile to intrinsics on modern compiler for short copies. However my argument is kind of weak since modern compilers will also recognize our hand written memcpy/memset loops for example. If a platform misses those, simple versions could be provided in like 5 lines of C. The only problem is that we are probably very often not doing bytewise copy, but wordwise. This is were we would get worse with a naive C implementation. But I think memset/memcpy will so or so being present in a binary since other parts of the code are using it. Then we are doing worse here since we generate bloat which would otherwise be unnecessary given a primitive compiler. So what remains in the end is maybe that these functions clearly show the intend (as clearly as a loop?).

* strcmp: Same with strcmp, should not be avoided. Also intrinsic.

* strcpy, strncpy: On these I agree. These are bad functions. They shouldn't be used anywhere.

* clock: Should be avoided obviously, this is a syscall.

* qsort: I agree on that one.

So in my opinion not much remains of the "unwanted function" list :wink:

Reply:
It seems you removed s_sqrt and simplified things a lot. I guess I am more happy now :)